### PR TITLE
[Sprint 22 / PD-50] - [Feature] Add implementation GRPC 

### PR DESCRIPTION
**Pull request by marfanr**
## Jira Link: 
https://ichiro-its.atlassian.net/browse/PD-50?atlOrigin=eyJpIjoiYzAxZWU5MzYxMGQwNDQ0ZWE3MjY2ZDQxMDg2MzdhYmEiLCJwIjoiaiJ9

## Description

added protobuf config for grpc

## Type of Change

- [ ] Bugfix
- [ ] Enhancement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause the existing functionality to not work as expected)

## How Has This Been Tested?

- [ ] New unit tests added.
- [x] Manual tested.

## Checklist:

- [x] Using Branch Name Convention
    - `feature/JIRA-ID-SHORT-DESCRIPTION` if has a JIRA ticket
    - `enhancement/SHORT-DESCRIPTION` if has/has no JIRA ticket and contain enhancement
    - `hotfix/SHORT-DESCRIPTION` if the change doesn't need to be tested (urgent)
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made the documentation for the corresponding changes.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -16,6 +16,47 @@ endif()
...
+add_library(aruku_proto
+  ${${CMAKE_CURRENT_BINARY_DIR}/aruku.pb.cc}
+  ${${CMAKE_CURRENT_BINARY_DIR}/aruku.pb.h}
+  ${${CMAKE_CURRENT_BINARY_DIR}/aruku.pb.h}
+  ${${CMAKE_CURRENT_BINARY_DIR}/aruku.grpc.pb.h})
+
+target_link_libraries(aruku_proto
+  ${_REFLECTION}
+  ${_GRPC_GRPCPP}
+  ${_PROTOBUF_LIBPROTOBUF}
```
Comment (threeal) [inline CMakeLists.txt]: We can just put the libraries here:
```suggestion
  gRPC::grpc++_reflection
  gRPC::grpc++
  protobuf::libprotobuf
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -16,6 +16,47 @@ endif()
...
+set(_GRPC_CPP_PLUGIN_EXECUTABLE $<TARGET_FILE:gRPC::grpc_cpp_plugin>)
+
+# Proto file
+get_filename_component(aruku_proto "./proto/aruku.proto" ABSOLUTE)
+get_filename_component(aruku_proto_path "${aruku_proto}" PATH)
+
+# Generated sources
+add_custom_command(
+      OUTPUT "${${CMAKE_CURRENT_BINARY_DIR}/aruku.pb.cc}" "${${CMAKE_CURRENT_BINARY_DIR}/aruku.pb.h}" "${${CMAKE_CURRENT_BINARY_DIR}/aruku.pb.h}" "${${CMAKE_CURRENT_BINARY_DIR}/aruku.grpc.pb.h}"
+      COMMAND ${_PROTOBUF_PROTOC}
```
Comment (threeal) [inline CMakeLists.txt]: We can just use `$<TARGET_FILE:protobuf::protoc>` here:
```suggestion
      COMMAND $<TARGET_FILE:protobuf::protoc>
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -16,6 +16,47 @@ endif()
...
+      ARGS --grpc_out "${CMAKE_CURRENT_BINARY_DIR}"
+        --cpp_out "${CMAKE_CURRENT_BINARY_DIR}"
+        -I "${aruku_proto_path}"
+        --plugin=protoc-gen-grpc="${_GRPC_CPP_PLUGIN_EXECUTABLE}"
```
Comment (threeal) [inline CMakeLists.txt]: We can just use `$<TARGET_FILE:gRPC::grpc_cpp_plugin>` in here:
```suggestion
        --plugin=protoc-gen-grpc=$<TARGET_FILE:gRPC::grpc_cpp_plugin>
```

File: CMakeLists.txt [hunk 4]
```diff
@@ -16,6 +16,47 @@ endif()
...
+set(_PROTOBUF_PROTOC $<TARGET_FILE:protobuf::protoc>)
+
+find_package(gRPC CONFIG REQUIRED)
+message(STATUS "Using gRPC ${gRPC_VERSION}")
+
+set(_GRPC_GRPCPP gRPC::grpc++)
+set(_GRPC_CPP_PLUGIN_EXECUTABLE $<TARGET_FILE:gRPC::grpc_cpp_plugin>)
+
+# Proto file
+get_filename_component(aruku_proto "./proto/aruku.proto" ABSOLUTE)
```
Comment (threeal) [inline CMakeLists.txt]: Much accurate:
```suggestion
set(aruku_proto ${CMAKE_CURRENT_SOURCE_DIR}/proto/aruku.proto)
```

File: CMakeLists.txt [hunk 5]
```diff
@@ -16,6 +16,47 @@ endif()
...
+get_filename_component(aruku_proto_path "${aruku_proto}" PATH)
```
Comment (threeal) [inline CMakeLists.txt]: Please use `DIRECTORY`, see [this](https://cmake.org/cmake/help/latest/command/get_filename_component.html).
```suggestion
get_filename_component(aruku_proto_path "${aruku_proto}" DIRECTORY)
```

File: CMakeLists.txt [hunk 6]
```diff
@@ -16,6 +16,47 @@ endif()
...
+        -I "${aruku_proto_path}"
+        --plugin=protoc-gen-grpc="${_GRPC_CPP_PLUGIN_EXECUTABLE}"
+        "${aruku_proto}"
+      DEPENDS "${aruku_proto}")
+
+add_library(aruku_proto
+  ${${CMAKE_CURRENT_BINARY_DIR}/aruku.pb.cc}
+  ${${CMAKE_CURRENT_BINARY_DIR}/aruku.pb.h}
+  ${${CMAKE_CURRENT_BINARY_DIR}/aruku.pb.h}
+  ${${CMAKE_CURRENT_BINARY_DIR}/aruku.grpc.pb.h})
```
Comment (threeal) [inline CMakeLists.txt]: Why there's double `${}` in here? Won't it work?

```suggestion
  ${CMAKE_CURRENT_BINARY_DIR}/aruku.pb.cc
  ${CMAKE_CURRENT_BINARY_DIR}/aruku.pb.h
  ${CMAKE_CURRENT_BINARY_DIR}/aruku.pb.h
  ${CMAKE_CURRENT_BINARY_DIR}/aruku.grpc.pb.h)
```

**Review by threeal — CHANGES_REQUESTED** [review]

**Review by threeal — APPROVED** [review]

**Review by FaaizHaikal — APPROVED** [review]

**Pull request merged**